Add persisted and new_records methods to ActiveRecord::Associations::CollectionProxy.

stevepolitodesign opened the pull request:
This is a work in progress, and I wanted to get some feedback before I go too far.

This aims to extend the [methods added by has_many](https://guides.rubyonrails.org/association_basics.html#methods-added-by-has-many) and account for cases where a record is not yet saved but is still added to a collection.

## Usage

```ruby
class Post < ApplicationRecord
  has_many :comments
end

Post.first.comments.persisted
```

## Example

### Before

This will render all saved comments as well as the newly initialized comment.

```ruby
class PostsController < ApplicationController
  def show
    @post = Post.find(params[:id])
    @comment = @post.comments.build
    @comments = @post.comments
  end
end
```

```html+erb
<%= render @comments %>
```

### After

This will only render comments that have been saved to the database.

```ruby
class PostsController < ApplicationController
  def show
    @post = Post.find(params[:id])
    @comment = @post.comments.build
    @comments = @post.comments.persisted
  end
end
```

```html+erb
<%= render @comments %>
```

## To Do

- [x] Add and test `new_records` method.
- [x] Update changelong
- [x] Update the Guides
- [x] Update the API

<!-- If there's anything else that's important and relevant to your pull
request, mention that information here. This could include
benchmarks, or other information.

If you are updating any of the CHANGELOG files or are asked to update the
CHANGELOG files by reviewers, please add the CHANGELOG entry at the top of the file.

Finally, if your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

Thanks for contributing to Rails! -->

<!--
Note: Please avoid making *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.
Create a pull request when it is ready for review and feedback
from the Rails team :).
-->

stevepolitodesign committed:
Add persisted and new_records methods to ActiveRecord::Associations::CollectionProxy.

Prior to this change there was no way to distinguish between persisted records and new records when returning a collection.

```ruby
class Person < ActiveRecord::Base
  has_many :pets
end

person.pets.create(name: 'Fancy-Fancy')
=> #<Pet id: 1, name: "Fancy-Fancy", person_id: 1>

person.pets.build([{name: 'Spook'}, {name: 'Choo-Choo'}])
=> [
  <Pet id: nil, name: "Spook", person_id: 1>,
  <Pet id: nil, name: "Choo-Choo", person_id: 1>
]

person.pets
=> [
  <Pet id: 1, name: "Fancy-Fancy", person_id: 1>,
  <Pet id: nil, name: "Spook", person_id: 1>,
  <Pet id: nil, name: "Choo-Choo", person_id: 1>
]
```

After this change you can call `persisted` or `new_records` on the collection.

```ruby
class Person < ActiveRecord::Base
  has_many :pets
end

person.pets.create(name: 'Fancy-Fancy')
=> #<Pet id: 1, name: "Fancy-Fancy", person_id: 1>

person.pets.build([{name: 'Spook'}, {name: 'Choo-Choo'}])
=> [
  #<Pet id: nil, name: "Spook", person_id: 1>,
  #<Pet id: nil, name: "Choo-Choo", person_id: 1>
]

person.pets.persisted
=> [#<Pet id: 1, name: "Fancy-Fancy", person_id: 1>]

person.pets.new_records
=> [
  #<Pet id: nil, name: "Spook", person_id: 1>,
  #<Pet id: nil, name: "Choo-Choo", person_id: 1>
]
```

excid3 commented on activerecord/lib/active_record/associations/collection_proxy.rb:
```diff
@@ -1091,6 +1091,11 @@ def reset
 
+      def persisted
+        primary_key_column = @association.reflection.active_record_primary_key
+        load_target.select { |element| element[primary_key_column] != nil }
```
Would you be able to use the `persisted?` method instead?

```ruby
def persisted
  load_target.select(&:persisted?)
end
```

DRBragg commented:
I dig this!

excid3 commented:
This is an area of Rails the catches a lot of people off guard, so I think this would be a useful method.

ghiculescu commented:
I like this, did you consider the inverse (`comments.new_records`) too?

Please just remember to squash + add a changelog entry when you're happy with this.

stevepolitodesign commented:
@ghiculescu I went ahead and addressed the following:

- Added a `new_records` method
- Updated the `CHANGELOG.md`
- Documented the methods in `activerecord/lib/active_record/associations/collection_proxy.rb`
- Squashed my commits.

**Questions**

Should I also update the guides since this would affect [4.3.1 Methods Added by has_many](https://guides.rubyonrails.org/association_basics.html#methods-added-by-has-many) and [4.4.1 Methods Added by has_and_belongs_to_many](https://guides.rubyonrails.org/association_basics.html#methods-added-by-has-and-belongs-to-many)

ghiculescu commented:
> Should I also update the guides since this would affect [4.3.1 Methods Added by has_many](https://guides.rubyonrails.org/association_basics.html#methods-added-by-has-many) and [4.4.1 Methods Added by has_and_belongs_to_many](https://guides.rubyonrails.org/association_basics.html#methods-added-by-has-and-belongs-to-many)

Yeah I think so.

stevepolitodesign committed:
Add new methods to guides.

stevepolitodesign committed:
Credit Chris since he reviewed the PR.

stevepolitodesign committed:
Merge branch 'main' into has_many_association_persisted_records

stevepolitodesign commented:
Just a heads up, I'm currently unable to squash these into one commit due to some merge conflicts. I wouldn't want that to prevent this from being merged in, but I also don't know enough about git to resolve this at the moment.

stevepolitodesign commented:
@rafaelfranca thank you for taking the time to review this. 

> ... I'm still uneasy to add those methods that looks like scopes and we would think they filter records in the database, but in reality they are filtering the records in memory.

Would it be better to create `persisted_collection` and/or `new_collection` methods? 

```ruby
book.persisted_authors
book.new_authors
```

I'm not sure if the `new_collection` method would have much value, but I think having a way to deliberately return persisted records is still valuable.

If you still have reservations about this proposal, maybe we should consider updating the guides to make developers aware that calling `collection` will return both persisted and newly instantiated records as I don't think this is made clear.

ghiculescu commented:
Thinking out loud, could the new methods behave differently based on `loaded?`. For example, for `#persisted`:

```ruby
def persisted
  if loaded?
    load_target.select(&:persisted?)
  else
    # this isn't quite right, since you'd also want anything persisted `post.comments.create`, but something along these lines...
    all
  end
end
```
